### PR TITLE
Discover dacs

### PR DESCRIPTION
**Pull request by pkivolowitz**
It's a little messy - but effective.

### CODE REVIEW

File: curses/client.cpp [hunk 1]
```diff
@@ -507,14 +502,14 @@ int main(int argc, char * argv[])
 				{
 					case '+':
 						dac_number++;
-						if (dac_number >= MAX_DACS)
+						if (dac_number >= number_of_dacs)
```
Comment (LOZORD) [inline curses/client.cpp]: This is getting to be a little too deeply nested. Can you pull a few of these scopes into functions?
Also, having two variables called `dac_number` and `number_of_dacs` might get confusing. Is there a way to calculate these on the fly? For example is `number_of_dacs == dac_list.length()` (assuming such list exists)?

File: server/audio_device.hpp [hunk 2]
```diff
@@ -24,15 +24,17 @@ struct AudioDevice
 	{
 	}
 
-	AudioDevice(std::string s, std::string n, int i)
+	AudioDevice(std::string s, std::string n, std::string a, int i)
```
Comment (LOZORD) [inline server/audio_device.hpp]: Depending on what style you prefer, an [initialization list](http://en.cppreference.com/w/cpp/language/initializer_list) might be better here.

File: server/discover_dacs.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	FILE* p = nullptr;
+	string cmdline = string("pacmd list-sinks | grep name");
+
+	if ((p = popen(cmdline.c_str(), "r")) == nullptr) {
+		LOG2(_log_, string("pipe to pacmd failed: ") + string(strerror(errno)), FATAL);
+		return devices;
+	}
+
+	char buffer[BSIZE] = { 0 };
+	string device_name = "device.product.name";
```
Comment (LOZORD) [inline server/discover_dacs.cpp]: Make this a constant?

File: server/makefile [hunk 4]
```diff
@@ -5,7 +5,7 @@ CFLAGS	= -g -O3 -Wall -std=c++11 -pthread -D _DEBUG_
 LDFLAGS	= -lavcodec -lavformat -lmysqlcppconn -lpulse-simple -lpulse -lrt -lprotobuf
 
 SRC		= server.cpp logger.cpp connection_manager.cpp network_component.cpp \
-		  db_component.cpp track.cpp audio_component.cpp \
+		  db_component.cpp track.cpp audio_component.cpp discover_dacs.cpp \
```
Comment (LOZORD) [inline server/makefile]: Might be good to list one `cpp` file per line for easier addition/removal. Up to you though.

File: server/server.cpp [hunk 5]
```diff
@@ -67,12 +68,12 @@ int main(int argc, char * argv[])
 	
 	try
 	{
-		vector<string> valid_extensions;
-
-		devices.push_back(AudioDevice("alsa_output.usb-AudioQuest_AudioQuest_DragonFly_Black_v1.5_AQDFBL0100111808-01.analog-stereo", "dragonFly Black", (int) devices.size()));
-		devices.push_back(AudioDevice("alsa_output.usb-Audioengine_Audioengine_D3_Audioengine-00.analog-stereo", "audioengine D3", (int) devices.size()));
-		devices.push_back(AudioDevice("alsa_output.usb-FiiO_DigiHug_USB_Audio-01.analog-stereo", "Fiio", (int) devices.size()));
+		vector<string> valid_extensions = { "mp3", "flac", "wav", "m4a" ,"ogg" };
```
Comment (LOZORD) [inline server/server.cpp:71]: Don't need this _in_ the try block.

**Review by LOZORD — COMMENTED** [review]
Looks good so far. Next steps should be to remove all curses-related code from this repo and move it into another one.

File: curses/client.cpp (hunk 1 above)
Comment (pkivolowitz) [reply curses/client.cpp]: The list doesn't exist but renaming dac_number to current_dac_index should help.

The contents of the big switch statement has been cleaned up.

Comment (LOZORD) [reply curses/client.cpp]: Awesome, thanks.

File: server/audio_device.hpp (hunk 2 above)
Comment (pkivolowitz) [reply server/audio_device.hpp]: Shifted all appropriate constructors to initializers.

**Comment by pkivolowitz** [comment]
As I make these changes, what do I do with the pull request?

**Pull request merged**